habitat_code/kobylarz_kmeans_review_code_20240313.py

```diff
@@ -0,0 +1,603 @@
+# 
+# **YOUR TASK:**
+# 1. For each granule:
+#     1. Open the Fmask band, crop, and compute a quality mask for the granule. You can use the following code as a starting point, making sure that `mask_bits` contains the quality bits you want to consider:
+#          ```python
+#          # Expand into a new dimension of binary bits
+#          bits = np.unpackbits(da.astype(np.uint8)).reshape(da.shape + (-1,))
+# 
+#          # Select the required bits and check if any are flagged
+#          mask = np.prod(bits[..., mask_bits]==0, axis=-1)
+#          ```
+# 
+#     2. For each band that starts with 'B':
+#         1. Open the band, crop, and apply the scale factor
+#         2. Name the DataArray after the band using the `.name` attribute
+#         3. Apply the cloud mask using the `.where()` method
+#         4. Store the DataArray in your data structure (e.g. adding a GeoDataFrame column with the DataArray in it. Note that you will need to remove the rows for unused bands)
+
+# In[9]:
+
+
+# Function to compute image mask
+def process_tifs(download_df, huc_gdf):
+    """
+    Load and process the Fmask and B band .tif files
+    in a granule.
+
+    Parameters
+    ==========
+    download_df : pandas.core.groupby.generic.DataFrame
+      Dataframe containing all the rows for each band .tif for all the files.
+    
+    huc_gdf : GeoDataFrame
+      Geodataframe containing the geometry of the HUC watershed used to clip
+
+    Returns
+    =======
+    accumulator_df : Pandas dataframe
+      Dataframe containing the processed rows for all the tifs.
+      An xarray has been added for each processed .tif in the last column.
+    """
+    grouped = download_df.groupby('granule_id')
+    columns = ['geometry',
+               'granule_id', 
+               'file_url', 
+               'tile_id', 
+               'datetime', 
+               'band_number',
+               'processed_da']
+    accumulator_df = pd.DataFrame(columns=columns)
+    # Looping through each Granule and processing data
+    # Create cloud mask
+    for granule_id, group_data in grouped:
+        print("Processing data for Granule ID:", granule_id)
+        granule_gdf = group_data
+        fmask_gdf = group_data['band_number'] == "Fmask"
+        fmask_gdf = group_data[fmask_gdf].iloc[0:1]
+        file_name = (
+            fmask_gdf.iloc[0]['granule_id'] +
+            "." + fmask_gdf.iloc[0]['band_number'] + '.tif'
+        )
+        file_path = data_dir + '\\earthpy-downloads\\' + file_name
+        fmask_da = rxr.open_rasterio(file_path).squeeze()
+        huc_gdf_crs = huc_gdf.to_crs(fmask_da.rio.crs)
+        fmask_crop_da = fmask_da.rio.clip_box(*huc_gdf_crs.total_bounds)
+        mask = compute_mask(fmask_crop_da)
+        # Apply crop, scale factor, and mask to all "B" files
+        for index, row in group_data.iterrows():
+            if (row['band_number'].startswith('B')
+                and row['band_number'] not in ('B10', 'B11')):
```

Comment:
I recommend indenting this line at 6 or 8 spaces to distinguish it from the inside of the conditional.